File: chapter2/sort_of_selection_sort/1.go
package main

import "fmt"

func findSmallest(arr []int) int {
	smallestIndex := 0
	smallest := arr[smallestIndex]

	arrLength := len(arr)

	for i := 1; i < arrLength; i++ {
		if arr[i] < smallest {
			smallest = arr[i]
			smallestIndex = i
		}
	}

	return smallestIndex
}

func selectionSort(arr []int) []int {
	arrLength := len(arr)

	sortedArr := make([]int, arrLength)

	for i := 0; i < arrLength; i++ {
		smallestIndex := findSmallest(arr)
		sortedArr[i] = arr[smallestIndex]

		arr = append(arr[:smallestIndex], arr[smallestIndex+1:]...)
	}

	return sortedArr
}

func main() {
	fmt.Println(selectionSort([]int{5, 3, 6, 2, 10}))
}

File: chapter2/sort_of_selection_sort/1.js
const findSmallest = (arr) => {
  let smallestIndex = 0;
  let smallest = arr[smallestIndex];

  const arrLength = arr.length;

  for (let i = 1; i < arrLength; i++) {
    if (arr[i] < smallest) {
      smallest = arr[i];
      smallestIndex = i;
    }
  }

  return smallestIndex
};

const selectionSort = (arr) => {
  sortedArr = [];

  const arrLength = arr.length;

  for (let i = 0; i < arrLength; i++) {
    smallestIndex = findSmallest(arr);

    sortedArr.push(...arr.splice(smallestIndex, 1));
  }

  return sortedArr
};

console.log(selectionSort([5, 3, 6, 2, 10]))